apt_info.py

```diff
@@ -57,7 +57,7 @@ def _write_pending_upgrades(registry, cache):
     # only one upgrade, not two). See the following issue for more details:
     # https://github.com/prometheus-community/node-exporter-textfile-collector-scripts/issues/85
     candidates = {
-        p.candidate for p in cache.get_changes() if p.is_installed and p.marked_upgrade
+        p.candidate for p in cache if p.is_installed and p.is_upgradable
```

Comment:
Isn't it sufficient to simply test for `p.is_gradable` (as is done in the `_write_held_upgrades` function)? I'm pretty sure that a package won't be "upgradable" if it is not already "installed".

Reply:
I believe that it was carried over from the previous code that tested `p.marked_upgrade`. It was only intended to be a quick commit fixing the issue.

I have done some more testing on my Proxmox VE 8.1 node which is yet to be upgraded to PVE 8.2 to investigate:

```python3
>>> import apt
>>> import apt_pkg
>>> cache = apt.cache.Cache()
>>> # the following calls on cache including cache.upgrade(True) were removed in #181 which makes counting marked_upgrade not working.
>>> # In this session we run those again so we can compare the results in the old and new implementation
>>> cache.update()
True
>>> cache.open()
>>> cache.upgrade(True)
>>> sorted([p.name for p in cache if p.is_installed and p.is_upgradable])
['grub-common', 'grub-efi-amd64-bin', 'grub-pc', 'grub-pc-bin', 'grub2-common', 'libc-bin', 'libc-l10n', 'libc6', 'libpve-access-control', 'libpve-cluster-api-perl', 'libpve-cluster-perl', 'libpve-common-perl', 'libpve-guest-common-perl', 'libpve-http-server-perl', 'libpve-network-perl', 'libpve-notify-perl', 'libpve-storage-perl', 'locales', 'proxmox-backup-client', 'proxmox-backup-file-restore', 'proxmox-backup-restore-image', 'proxmox-default-kernel', 'proxmox-ve', 'proxmox-widget-toolkit', 'pve-cluster', 'pve-container', 'pve-docs', 'pve-firewall', 'pve-ha-manager', 'pve-i18n', 'pve-manager', 'pve-qemu-kvm', 'qemu-server']
>>> sorted([p.name for p in cache if p.is_upgradable])
['grub-common', 'grub-efi-amd64-bin', 'grub-pc', 'grub-pc-bin', 'grub2-common', 'libc-bin', 'libc-l10n', 'libc6', 'libpve-access-control', 'libpve-cluster-api-perl', 'libpve-cluster-perl', 'libpve-common-perl', 'libpve-guest-common-perl', 'libpve-http-server-perl', 'libpve-network-perl', 'libpve-notify-perl', 'libpve-storage-perl', 'locales', 'proxmox-backup-client', 'proxmox-backup-file-restore', 'proxmox-backup-restore-image', 'proxmox-default-kernel', 'proxmox-ve', 'proxmox-widget-toolkit', 'pve-cluster', 'pve-container', 'pve-docs', 'pve-firewall', 'pve-ha-manager', 'pve-i18n', 'pve-manager', 'pve-qemu-kvm', 'qemu-server']
>>> sorted([p.name for p in cache.get_changes() if p.is_installed and p.marked_upgrade])
['grub-common', 'grub-efi-amd64-bin', 'grub-pc', 'grub-pc-bin', 'grub2-common', 'libc-bin', 'libc-l10n', 'libc6', 'libpve-access-control', 'libpve-cluster-api-perl', 'libpve-cluster-perl', 'libpve-common-perl', 'libpve-guest-common-perl', 'libpve-http-server-perl', 'libpve-network-perl', 'libpve-notify-perl', 'libpve-storage-perl', 'locales', 'proxmox-backup-client', 'proxmox-backup-file-restore', 'proxmox-backup-restore-image', 'proxmox-default-kernel', 'proxmox-ve', 'proxmox-widget-toolkit', 'pve-cluster', 'pve-container', 'pve-docs', 'pve-firewall', 'pve-ha-manager', 'pve-i18n', 'pve-manager', 'pve-qemu-kvm', 'qemu-server']
>>> sorted([p.name for p in cache.get_changes() if p.marked_upgrade])
['grub-common', 'grub-efi-amd64-bin', 'grub-pc', 'grub-pc-bin', 'grub2-common', 'libc-bin', 'libc-l10n', 'libc6', 'libpve-access-control', 'libpve-cluster-api-perl', 'libpve-cluster-perl', 'libpve-common-perl', 'libpve-guest-common-perl', 'libpve-http-server-perl', 'libpve-network-perl', 'libpve-notify-perl', 'libpve-storage-perl', 'locales', 'proxmox-backup-client', 'proxmox-backup-file-restore', 'proxmox-backup-restore-image', 'proxmox-default-kernel', 'proxmox-firewall', 'proxmox-kernel-6.8', 'proxmox-kernel-6.8.4-2-pve-signed', 'proxmox-ve', 'proxmox-widget-toolkit', 'pve-cluster', 'pve-container', 'pve-docs', 'pve-firewall', 'pve-ha-manager', 'pve-i18n', 'pve-manager', 'pve-qemu-kvm', 'qemu-server']
>>> set(p.name for p in cache.get_changes() if p.marked_upgrade) - set(p.name for p in cache.get_changes() if p.is_installed and p.marked_upgrade)
{'proxmox-firewall', 'proxmox-kernel-6.8.4-2-pve-signed', 'proxmox-kernel-6.8'}
>>> cache.close()
```

You can see that with the old implementation, `p.marked_upgrade` would include new packages that would be installed due to changed dependency in upgraded packages (for example `proxmox-kernel-6.8.4-2-pve-signed` due to an upgraded kernel). Therefore in the old implementation, `p.is_installed` is also checked to discount any changes that apply to packages that aren't installed).

However, with my new implementation, there is no difference between `p.is_installed and p.is_upgradable` and `p.is_upgradable`. I believe that it is sufficient to test `p.is_upgradable` alone. I'll modify the code accordingly

Reply:
Fixed in 51bb32c12a1b8195a327b31c694875f2a86e5684